src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -657,9 +479,359 @@ protected virtual void NotifyListeners(ProjectChangeEventArgs e)
         }
     }
 
+    internal override void UpdateProject(
+        string filePath,
+        RazorConfiguration configuration,
+        ProjectWorkspaceState projectWorkspaceState,
+        string? rootNamespace,
+        Func<IProjectSnapshot, ImmutableArray<IUpdateProjectAction>> calculate)
+    {
+        if (projectWorkspaceState is null)
+        {
+            throw new ArgumentNullException(nameof(projectWorkspaceState));
+        }
+
+        using var _ = ListPool<ProjectChangeEventArgs>.GetPooledObject(out var changesToNotify);
+
+        // Get an upgradeableLock, which will keep a read lock while we compute the changes
+        // and then get a write lock to actually apply them. Only one upgradeable lock
+        // can be held at any given time. Write lock must be retrieved on the same
+        // thread that the lock was acquired
+        using (var upgradeableLock = _locksFactory.EnterUpgradeAbleReadLock())
+        {
+            UpdateProject_NoLock(filePath, configuration, projectWorkspaceState, rootNamespace, calculate, changesToNotify, upgradeableLock);
+        }
+
+        // Notify outside of the lock, since notifications may trigger mutations from listeners
+        foreach (var notification in changesToNotify)
+        {
+            NotifyListeners(notification);
+        }
+    }
+
+    private void UpdateProject_NoLock(
+        string filePath,
+        RazorConfiguration configuration,
+        ProjectWorkspaceState projectWorkspaceState,
+        string? rootNamespace,
+        Func<IProjectSnapshot, ImmutableArray<IUpdateProjectAction>> calculate,
+        List<ProjectChangeEventArgs> changesToNotify,
+        in ReadWriterLocker.UpgradeableReadLock upgradeableLock)
```

Comment:
Consider passing this as `ref`, since `UpgradeableReadLock` is mutable. That'll ensure that a subtle bug won't be introduced if this method ever starts disposing `upgradeableLock` and the caller thinks it hasn't been disposed because the internal state of the struct isn't copied back to the caller.

Reply:
Interestingly, I did that first but found that you can't use `ref` on variables that are using variables. https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs1657?f1url=%3FappId%3Droslyn%26k%3Dk(CS1657)

Reply:
Ah, right. Yeah, you'd have to go with a try/finally. It'd be more correct, but it might not be worth it in hindsight.